contracts/rsv/Reserve.sol

```diff
@@ -0,0 +1,367 @@
+
+    // Auth roles
+    address public minter;
+    address public pauser;
+    address public freezer;
+    address public feeRecipient;
+
+    // Auth role change events
+    event MinterChanged(address indexed newMinter);
+    event PauserChanged(address indexed newPauser);
+    event FreezerChanged(address indexed newFreezer);
+    event FeeRecipientChanged(address indexed newFeeRecipient);
+    event MaxSupplyChanged(uint256 indexed newMaxSupply);
+
+    // Pause events
+    event Paused(address indexed account);
+    event Unpaused(address indexed account);
+
+    // Name change event
+    event NameChanged(string newName, string newSymbol);
+
+    // Law enforcement events
+    event Frozen(address indexed freezer, address indexed account);
+    event Unfrozen(address indexed freezer, address indexed account);
+    event Wiped(address indexed freezer, address indexed wiped);
+
+
+    /// Initialize critical fields.
+    constructor() public {
+        data = new ReserveEternalStorage(msg.sender);
+        txFee = ITXFee(address(0)); // I'm not sure if this will fail here or later. If it fails here, then we'll need a different design
+        pauser = msg.sender;
+        feeRecipient = msg.sender;
+        maxSupply = 2 ** 256 - 1;
+        // Other roles deliberately default to the zero address.
+    }
+
+    /// Accessor for eternal storage contract address.
+    function getEternalStorageAddress() external view returns(address) {
+        return address(data);
+    }
+
+
+    // ==== Admin functions ====
+
+
+    /// Modifies a function to only run if sent by `role`.
+    modifier only(address role) {
+        require(msg.sender == role, "unauthorized: not role holder");
+        _;
+    }
+
+    /// Modifies a function to only run if sent by `role` or the contract's `owner`.
+    modifier onlyOwnerOr(address role) {
+        require(msg.sender == _owner || msg.sender == role, "unauthorized: not role holder and not owner");
+        _;
+    }
+
+    /// Change who holds the `minter` role.
+    function changeMinter(address newMinter) external onlyOwnerOr(minter) {
+        minter = newMinter;
+        emit MinterChanged(newMinter);
+    }
+
+    /// Change who holds the `pauser` role.
+    function changePauser(address newPauser) external onlyOwnerOr(pauser) {
+        pauser = newPauser;
+        emit PauserChanged(newPauser);
+    }
+
+    /// Change who holds the `freezer` role.
+    function changeFreezer(address newFreezer) external onlyOwnerOr(freezer) {
+        freezer = newFreezer;
+        emit FreezerChanged(newFreezer);
+    }
+
+    function changeFeeRecipient(address newFeeRecipient) external onlyOwnerOr(feeRecipient) {
+        feeRecipient = newFeeRecipient;
+        emit FeeRecipientChanged(newFeeRecipient);
+    }
+
+    /// Make a different address own the EternalStorage contract.
+    /// This will break this contract, so only do it if you're
+    /// abandoning this contract, e.g., for an upgrade.
```

Comment:
Perhaps we can add an upgrade bool to set and a require here as an extra precaution? Maybe overkill / not adding value though. Not sure.